File: tips.js
function loadNouns(filename) {
    // Simulating an asynchronous file read operation
    return fetch(filename)
        .then(response => response.text())
        .then(data => data.split('\n').map(line => line.trim()));
}

function containsNoChars(noun, no) {
    return no.some(char => noun.includes(char));
}

function doesNotMatchExact(noun, exact) {
    for (let i = 0; i < 5; i++) {
        if (exact[i] !== '' && noun[i] !== exact[i]) {
            return true;
        }
    }
    return false;
}

function matchesNotThere(noun, notThere) {
    for (let i = 0; i < 5; i++) {
        for (const l of notThere[i]) {
            if (l !== '' && noun[i] === l) {
                return true;
            }
        }
    }
    return false;
}

function doesNotContainMustBe(noun, mustBe) {
    Array.from(mustBe).map((char) => console.log(noun, char))
    return mustBe.size === 0 ? false : !Array.from(mustBe).every(char => noun.includes(char));
}

function filterNouns(nouns, no, exact, notThere) {
    let mustBe = new Set();

    for (let i = 0; i < 5; i++) {
        mustBe.add(...exact[i]);
        mustBe.add(...notThere[i].split(''));
    }

    mustBe = new Set([...mustBe].filter(item => item !== undefined));

    no = no.filter(char => !mustBe.has(char));

    const possible = [];
    for (const noun of nouns) {
        if (containsNoChars(noun, no)) continue;
        if (doesNotMatchExact(noun, exact)) continue;
        if (matchesNotThere(noun, notThere)) continue;
        if (doesNotContainMustBe(noun, mustBe)) continue;

        possible.push(noun);
    }

    return possible;
}

const filename = 'sorted_nouns.txt';
let no = [];
const exact = ['', '', '', '', ''];
const notThere = ['', '', '', '', ''];

loadNouns(filename)
    .then(nouns => {
        const possible = filterNouns(nouns, no, exact, notThere);
        console.log(possible.length);
        console.log(possible);
    })
    .catch(error => {
        console.error('Error loading nouns:', error);
    });

File: tips.py
#!/usr/bin/env python
# coding: utf-8

def load_nouns(filename):
    with open(filename, 'r', encoding='utf-8') as file:
        return [line.strip() for line in file]

def filter_nouns(nouns, no, exact, not_there):
    must_be = set()

    for i in range(5):
        must_be.update(set(exact[i]))
        must_be.update(set(not_there[i]))

    possible = []
    for noun in nouns:
        if not set(no).isdisjoint(noun):
            continue

        continue_other = False
        for i in range(5):
            if exact[i] != '' and noun[i] != exact[i]:
                continue_other = True

        if continue_other:
            continue

        continue_other = False
        for i in range(5):
            for l in not_there[i]:
                if l != '' and noun[i] == l:
                    continue_other = True

        if continue_other:
            continue

        if not all(char in noun for char in must_be):
            continue

        possible.append(noun)

    return possible

if __name__ == "__main__":
    filename = 'sorted_nouns.txt'
    no = 'слово'
    exact = ['', '', '', '', '']
    not_there = ['', '', '', '', '']

    nouns = load_nouns(filename)
    possible = filter_nouns(nouns, no, exact, not_there)

    print(len(possible))
    print(possible)
